src/dft.c

```diff
@@ -1,8 +1,26 @@
 #include "gsa.h"
 #include <assert.h>
+#include <stdio.h>
+
+void dfs_rec(tree t, dynarr a) {
+    if(t->left != NULL ) dfs_rec(t->left, a);
+    append(&a, t->val);
+    if(t->right != NULL) dfs_rec(t->right, a);
+}
+
+void dfs(tree t, dynarr a) {
+    tree cur = t;
+    while (cur->left != NULL && cur->left) {
+        append(&a, cur->val);
+        if(cur->left->val < cur->right->val) cur = cur->left;
+        else cur = cur->right;
+    }
```

Comment:
I'm confused here.

What is it you are testing for here?

```c
while (cur->left != NULL && cur->left) {
   ...
}
```

You explicitly test for `NULL` for `cur->left` and implicitly for `cur->left` right after. The pointer is going to be either `NULL` or not, so `while (cur->left != NULL && cur->left)` is just `while(cur->left)`.

Of course, since you do `cur->left->val`, `cur->left` can't be `NULL`, but you also acces `cur->right->val` and you haven't checked if `cur->right` is `NULL`, so it could potentially crash there.

Maybe you meant to check if both children were non-`NULL`, but for a traversal, that isn't relevant at all.

Then, checking the values in the children is also irrelevant for a traversal, so I have no idea what that is there for.

Are you just searching down as far as you have a node with both children, picking the child with the smallest value at each recursive step? What is that doing?